From patchwork Wed Apr 28 17:38:42 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Josef Bacik <josef@toxicpanda.com>
X-Patchwork-Id: 12229869
Return-Path: <linux-btrfs-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-16.8 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,INCLUDES_PATCH,
	MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,USER_AGENT_GIT autolearn=ham
	autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 1E3FFC433ED
	for <linux-btrfs@archiver.kernel.org>; Wed, 28 Apr 2021 17:39:35 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id DD826613C6
	for <linux-btrfs@archiver.kernel.org>; Wed, 28 Apr 2021 17:39:34 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S242915AbhD1RkO (ORCPT <rfc822;linux-btrfs@archiver.kernel.org>);
        Wed, 28 Apr 2021 13:40:14 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:48196 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S242695AbhD1Rjh (ORCPT
        <rfc822;linux-btrfs@vger.kernel.org>);
        Wed, 28 Apr 2021 13:39:37 -0400
Received: from mail-qt1-x833.google.com (mail-qt1-x833.google.com
 [IPv6:2607:f8b0:4864:20::833])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id 4F2BFC061574
        for <linux-btrfs@vger.kernel.org>;
 Wed, 28 Apr 2021 10:38:52 -0700 (PDT)
Received: by mail-qt1-x833.google.com with SMTP id q4so18919807qtn.5
        for <linux-btrfs@vger.kernel.org>;
 Wed, 28 Apr 2021 10:38:52 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=toxicpanda-com.20150623.gappssmtp.com; s=20150623;
        h=from:to:subject:date:message-id:in-reply-to:references:mime-version
         :content-transfer-encoding;
        bh=5Z+HlITea37yRj2KAzeqd5CgcmKX15/MW14kAP9zMXs=;
        b=yZNhnpHk4mc9ZHevVQ+X7tgeh94NYpyam4/Zij9ISBdvBG9xY6qD8bzXrWikL6wWKB
         QzLw5uwzwQyUQec/PFKa8/1XIijk0yljH0LR+36OghFBi23JsqHY9ADLvwH8V1TkxJ3J
         7Wk/MuD7fvd+WQ/O4/8inONpY5k0zebBRdKZwvHScUsSEVZLebedvn47HcImidRAfVTe
         otq21YRZaCZDKRfNvVGpEy0k4TgrsmeFTwR4pRvRkuurvWmT/Yz5Z3rqWpJUDG7v2zP/
         U19nkye5Ms4RqqaEGvwRUBZXkqEgBXP43yKT5Ga/mjgddFUIrQ8xN2PYwhWO7jfRa0Y4
         BNwg==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=5Z+HlITea37yRj2KAzeqd5CgcmKX15/MW14kAP9zMXs=;
        b=QVTfeQRF6xBw9+2L4UOnQFkZtDLf2oatgpx5wPTMZuAh6QubrCN9HsPjQ2o6z7q76G
         1JOqilAUcERE0j+W7Fb0ypg2ImXBG3uekGByiSA4DwSmrqWPrAxEvsnO+jOq70bTnGs6
         YrPS0Bg2SN8ZnOhAxRX2Id/qV8I7GTf7OwXGdOztLf/TqNj685tzS96PORqUrmhl2A8S
         ZIhKuV2NzuwoeRIvVRADIrhB3EmiQ7OA0+gft5pbi81LoMAPGRadbIkiLTJiQ7OW82he
         gc3utXiPNogz0CKFS0hkEKMRbpeMkTUWdqbVm5dXRSBlXeX+dQAtWLMzWa1f55tcFDmc
         UCJA==
X-Gm-Message-State: AOAM533n2It7oisc3bpNHI9HFL+cpKXHjR9DCezaIS1SU9utw6wzDIxK
        /Cc0GO0UQmKt22fXVPqiWIeZ0VweyHQbyg==
X-Google-Smtp-Source: 
 ABdhPJyTCHjBCdaIl8BMCJkj0L107ChUTFstY1ueICrnTdBQG8a69wb1gu/GPHakViwRao1SKkJ74A==
X-Received: by 2002:ac8:7612:: with SMTP id
 t18mr27973028qtq.102.1619631531244;
        Wed, 28 Apr 2021 10:38:51 -0700 (PDT)
Received: from localhost (cpe-174-109-172-136.nc.res.rr.com.
 [174.109.172.136])
        by smtp.gmail.com with ESMTPSA id
 191sm310599qkj.17.2021.04.28.10.38.50
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Wed, 28 Apr 2021 10:38:50 -0700 (PDT)
From: Josef Bacik <josef@toxicpanda.com>
To: linux-btrfs@vger.kernel.org, kernel-team@fb.com
Subject: [PATCH 1/7] btrfs: check worker before need_preemptive_reclaim
Date: Wed, 28 Apr 2021 13:38:42 -0400
Message-Id: 
 <949c99db63e3f5c61dc402deee9ceae3087abae3.1619631053.git.josef@toxicpanda.com>
X-Mailer: git-send-email 2.26.3
In-Reply-To: <cover.1619631053.git.josef@toxicpanda.com>
References: <cover.1619631053.git.josef@toxicpanda.com>
MIME-Version: 1.0
Precedence: bulk
List-ID: <linux-btrfs.vger.kernel.org>
X-Mailing-List: linux-btrfs@vger.kernel.org

need_preemptive_reclaim() does some calculations, which aren't heavy,
but if we're already running preemptive reclaim there's no reason to do
them at all, so re-order the checks so that we don't do the calculation
if we're already doing reclaim.

Signed-off-by: Josef Bacik <josef@toxicpanda.com>
Reviewed-by: Johannes Thumshirn <johannes.thumshirn@wdc.com>
---
 fs/btrfs/space-info.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/fs/btrfs/space-info.c b/fs/btrfs/space-info.c
index 2dc674b7c3b1..c9a5e003bcfa 100644
--- a/fs/btrfs/space-info.c
+++ b/fs/btrfs/space-info.c
@@ -1588,8 +1588,8 @@ static int __reserve_bytes(struct btrfs_fs_info *fs_info,
 		 * the async reclaim as we will panic.
 		 */
 		if (!test_bit(BTRFS_FS_LOG_RECOVERING, &fs_info->flags) &&
-		    need_preemptive_reclaim(fs_info, space_info) &&
-		    !work_busy(&fs_info->preempt_reclaim_work)) {
+		    !work_busy(&fs_info->preempt_reclaim_work) &&
+		    need_preemptive_reclaim(fs_info, space_info)) {
 			trace_btrfs_trigger_flush(fs_info, space_info->flags,
 						  orig_bytes, flush, "preempt");
 			queue_work(system_unbound_wq,

From patchwork Wed Apr 28 17:38:43 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Josef Bacik <josef@toxicpanda.com>
X-Patchwork-Id: 12229873
Return-Path: <linux-btrfs-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-16.8 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,INCLUDES_PATCH,
	MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,USER_AGENT_GIT autolearn=ham
	autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 21ED4C433ED
	for <linux-btrfs@archiver.kernel.org>; Wed, 28 Apr 2021 17:40:09 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id DD91B61456
	for <linux-btrfs@archiver.kernel.org>; Wed, 28 Apr 2021 17:40:08 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S242586AbhD1Rkw (ORCPT <rfc822;linux-btrfs@archiver.kernel.org>);
        Wed, 28 Apr 2021 13:40:52 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:48198 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S242714AbhD1Rji (ORCPT
        <rfc822;linux-btrfs@vger.kernel.org>);
        Wed, 28 Apr 2021 13:39:38 -0400
Received: from mail-qv1-xf35.google.com (mail-qv1-xf35.google.com
 [IPv6:2607:f8b0:4864:20::f35])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id AF731C0613ED
        for <linux-btrfs@vger.kernel.org>;
 Wed, 28 Apr 2021 10:38:53 -0700 (PDT)
Received: by mail-qv1-xf35.google.com with SMTP id h3so30412543qve.13
        for <linux-btrfs@vger.kernel.org>;
 Wed, 28 Apr 2021 10:38:53 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=toxicpanda-com.20150623.gappssmtp.com; s=20150623;
        h=from:to:subject:date:message-id:in-reply-to:references:mime-version
         :content-transfer-encoding;
        bh=/A9DqPHUROVRO+mHvLHokSiw4oLb8UUfubXeGPKsepQ=;
        b=k5r0aVlKnpF2F8R4vMXswh/dBpvZLlcn79H7lVrwWgpGIq9oVJV6vw4TGUYGTIRukP
         e7S9KqZglBxRw6ps9hRE+ubT/BoCiNGVHDi/RL/RMU69k5/t0Lbt/EkLvrX6ufE7qoXu
         c0n03d86jcI1E2AMJwM/1qZ90p3k+AgFW57tugdXen2GUkd0nLg7MVh7cdIe4A0gfOLs
         aROKxZDAw5bY8M2GaDSb5672xLU7Nj9sw0C+6iFCmUAvu4Rch0o2W05xwBdI7Oh3WYcV
         fqqnzWqACI3LbE/UXr/dJFBswQDnj+jLtppmv5zkUDupGXrG7iXqG7kfAey/pjXAtc64
         ZIGg==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=/A9DqPHUROVRO+mHvLHokSiw4oLb8UUfubXeGPKsepQ=;
        b=AWV2Ljz2b4l+cYEqPTRuuwRuZDpcR71y0PGo8Km6OINv3L60DDAQC8+KZEHvnCb3sG
         kNxqAM9SjvmcQo3iGWKhKdWnYLTBGRKahnQK63Txe2avk7e9M222dJjDBkALJO4CoEgi
         QBtiaB2lJJeNnJSEZN+c3HZgnLO3QWa0dW6JAXjlBLcFB+MRLDo3uAfIA1+kImk2oLXn
         5k9IEVMeCnlHNBnfdmM3Bpivi/c7af5/hlexOZ1rhiGdSahc/bpFxOHLAuwPkjAxsVmW
         2/eq63ijYAZ55LIiQBX09D12Lk5bZCTDGmQ6/Iy+/pshkW0DmAhf9yP2Q+FcqPz6Wwwo
         oeHw==
X-Gm-Message-State: AOAM531lW833UG0i40Ig3dBAfQSpFPWcH/gENbfR2OpxURgyplvOgRH6
        sG9d84cbLx73+HsrQYXz15o8dvUTFk1LpA==
X-Google-Smtp-Source: 
 ABdhPJxbhh3lszSzWi3gHfEXGqJ4xtNy/FQ8wwdyQ3aHH/220oWLb0Ycr9ibMhKC+MFE0NZUbS4msQ==
X-Received: by 2002:ad4:5889:: with SMTP id dz9mr3344012qvb.12.1619631532662;
        Wed, 28 Apr 2021 10:38:52 -0700 (PDT)
Received: from localhost (cpe-174-109-172-136.nc.res.rr.com.
 [174.109.172.136])
        by smtp.gmail.com with ESMTPSA id h8sm328369qkj.4.2021.04.28.10.38.52
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Wed, 28 Apr 2021 10:38:52 -0700 (PDT)
From: Josef Bacik <josef@toxicpanda.com>
To: linux-btrfs@vger.kernel.org, kernel-team@fb.com
Subject: [PATCH 2/7] btrfs: only clamp the first time we have to start
 flushing
Date: Wed, 28 Apr 2021 13:38:43 -0400
Message-Id: 
 <03064e17bc0075187aed6330aceb77524aeadc3b.1619631053.git.josef@toxicpanda.com>
X-Mailer: git-send-email 2.26.3
In-Reply-To: <cover.1619631053.git.josef@toxicpanda.com>
References: <cover.1619631053.git.josef@toxicpanda.com>
MIME-Version: 1.0
Precedence: bulk
List-ID: <linux-btrfs.vger.kernel.org>
X-Mailing-List: linux-btrfs@vger.kernel.org

We were clamping the threshold for preemptive reclaim any time we added
a ticket to wait on, which if we have a lot of threads means we'd
essentially max out the clamp the first time we start to flush.  Instead
of doing this, simply do it every time we have to start flushing, this
will make us ramp up gradually instead of going to max clamping as soon
as we start needing to do flushing.

Signed-off-by: Josef Bacik <josef@toxicpanda.com>
---
 fs/btrfs/space-info.c | 17 +++++++++--------
 1 file changed, 9 insertions(+), 8 deletions(-)

diff --git a/fs/btrfs/space-info.c b/fs/btrfs/space-info.c
index c9a5e003bcfa..33edab17af0d 100644
--- a/fs/btrfs/space-info.c
+++ b/fs/btrfs/space-info.c
@@ -1561,6 +1561,15 @@ static int __reserve_bytes(struct btrfs_fs_info *fs_info,
 		    flush == BTRFS_RESERVE_FLUSH_DATA) {
 			list_add_tail(&ticket.list, &space_info->tickets);
 			if (!space_info->flush) {
+				/*
+				 * We were forced to add a reserve ticket, so
+				 * our preemptive flushing is unable to keep
+				 * up.  Clamp down on the threshold for the
+				 * preemptive flushing in order to keep up with
+				 * the workload.
+				 */
+				maybe_clamp_preempt(fs_info, space_info);
+
 				space_info->flush = 1;
 				trace_btrfs_trigger_flush(fs_info,
 							  space_info->flags,
@@ -1572,14 +1581,6 @@ static int __reserve_bytes(struct btrfs_fs_info *fs_info,
 			list_add_tail(&ticket.list,
 				      &space_info->priority_tickets);
 		}
-
-		/*
-		 * We were forced to add a reserve ticket, so our preemptive
-		 * flushing is unable to keep up.  Clamp down on the threshold
-		 * for the preemptive flushing in order to keep up with the
-		 * workload.
-		 */
-		maybe_clamp_preempt(fs_info, space_info);
 	} else if (!ret && space_info->flags & BTRFS_BLOCK_GROUP_METADATA) {
 		used += orig_bytes;
 		/*

From patchwork Wed Apr 28 17:38:44 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Josef Bacik <josef@toxicpanda.com>
X-Patchwork-Id: 12229883
Return-Path: <linux-btrfs-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-16.8 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,INCLUDES_PATCH,
	MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,USER_AGENT_GIT autolearn=ham
	autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 1177AC433B4
	for <linux-btrfs@archiver.kernel.org>; Wed, 28 Apr 2021 17:40:24 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id D3DC46144C
	for <linux-btrfs@archiver.kernel.org>; Wed, 28 Apr 2021 17:40:23 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S242683AbhD1RlC (ORCPT <rfc822;linux-btrfs@archiver.kernel.org>);
        Wed, 28 Apr 2021 13:41:02 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:48208 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S242744AbhD1Rjm (ORCPT
        <rfc822;linux-btrfs@vger.kernel.org>);
        Wed, 28 Apr 2021 13:39:42 -0400
Received: from mail-qt1-x834.google.com (mail-qt1-x834.google.com
 [IPv6:2607:f8b0:4864:20::834])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id 16C3DC061573
        for <linux-btrfs@vger.kernel.org>;
 Wed, 28 Apr 2021 10:38:55 -0700 (PDT)
Received: by mail-qt1-x834.google.com with SMTP id d12so9205600qtr.4
        for <linux-btrfs@vger.kernel.org>;
 Wed, 28 Apr 2021 10:38:55 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=toxicpanda-com.20150623.gappssmtp.com; s=20150623;
        h=from:to:subject:date:message-id:in-reply-to:references:mime-version
         :content-transfer-encoding;
        bh=qW1N7zEB+kpEjEKsVRxA1+eNSucsLWII2KirqoGlpPw=;
        b=RytIVD5jTepOB2igOAej1uDG9ieXMAoz0nryuF7W2AWaWaVC6ZkEgmuiT6+WQXnxqp
         T9Tuwc8OUAsUCo9YNA2jkGTAa6l2NE9F79cr+huKA1MZEm9UwlxjZ43UaPT82trSujFY
         7t2p2WNjuxDkqvXk/h5ITB/Kt91RnbrDGRsu3zmkkXBtS05WQrG3SJ2bPrkp0+XmCGTm
         2hLViBhSbcU8PAzAzzqtxY4INyV147ZHClJ6uIRz3FIWEVFd2Xi0mRndcM0/54U1DEHj
         0TmiMs9sZT7gxnMzyBX6tBZucfjdHlH4SAf9pDksxGkIaiTVHYSWUUlG6YQTRGjuCW87
         zipg==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=qW1N7zEB+kpEjEKsVRxA1+eNSucsLWII2KirqoGlpPw=;
        b=co0qIMcxi3uPSMwQQ0qROBeSSyEj9dF1f8XT8OB+johfTunEbcHc4I18m2LLK+h8Kq
         3J+MAfCojtdBGR5qgoAeq786ZcK8/TzTDYWrbvrAiD5Riv85L4/F+h05YSo0nFmTE3Rs
         9yFG3U6739ISN/JaaFtnraH90BP03mh+Qe4fIHsbIGyaU8cDu+IubBMBb/CX7sZck5nI
         MU1KzKryfFBjGbeZwElRmHssXYoVRj+NDm6mvleUk1uPHQeTAtSVfVGUNyPBXTNQUHqQ
         LzJICccG8T0EWBb0IHAwvl/dTJ/KKcQloWx0TEa42gzC5cM9zZ6t+vA35grn8HQhys7x
         Aa0g==
X-Gm-Message-State: AOAM530YpsKTsUqBjj5o49WbpnE2ZsbFkqqOZNoO+lZB+Q1tYTz09iIg
        dpd03HGfdhDCQLt3ZJc1gKNbrPmxKzgX2w==
X-Google-Smtp-Source: 
 ABdhPJzc7sUNgA9g5hRf2spxUyCz5I8jtnMmFaxnsnQUqPVUbaqedvuMe5EFgPuPF/IUzsDO7DA5zw==
X-Received: by 2002:ac8:6b19:: with SMTP id w25mr5917425qts.42.1619631534040;
        Wed, 28 Apr 2021 10:38:54 -0700 (PDT)
Received: from localhost (cpe-174-109-172-136.nc.res.rr.com.
 [174.109.172.136])
        by smtp.gmail.com with ESMTPSA id
 k18sm279866qkg.53.2021.04.28.10.38.53
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Wed, 28 Apr 2021 10:38:53 -0700 (PDT)
From: Josef Bacik <josef@toxicpanda.com>
To: linux-btrfs@vger.kernel.org, kernel-team@fb.com
Subject: [PATCH 3/7] btrfs: take into account global rsv in
 need_preemptive_reclaim
Date: Wed, 28 Apr 2021 13:38:44 -0400
Message-Id: 
 <febdf6d9ff9efc92034ed688d61b38a9b53a5abf.1619631053.git.josef@toxicpanda.com>
X-Mailer: git-send-email 2.26.3
In-Reply-To: <cover.1619631053.git.josef@toxicpanda.com>
References: <cover.1619631053.git.josef@toxicpanda.com>
MIME-Version: 1.0
Precedence: bulk
List-ID: <linux-btrfs.vger.kernel.org>
X-Mailing-List: linux-btrfs@vger.kernel.org

Global rsv can't be used for normal allocations, and for very full file
systems we can decide to try and async flush constantly even though
there's really not a lot of space to reclaim.  Deal with this by
including the global block rsv size in the "total used" calculation.

Signed-off-by: Josef Bacik <josef@toxicpanda.com>
Reviewed-by: Johannes Thumshirn <johannes.thumshirn@wdc.com>
---
 fs/btrfs/space-info.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/fs/btrfs/space-info.c b/fs/btrfs/space-info.c
index 33edab17af0d..e341f995a7dd 100644
--- a/fs/btrfs/space-info.c
+++ b/fs/btrfs/space-info.c
@@ -792,12 +792,14 @@
 static bool need_preemptive_reclaim(struct btrfs_fs_info *fs_info,
 				    struct btrfs_space_info *space_info)
 {
+	u64 global_rsv_size = fs_info->global_block_rsv.reserved;
 	u64 ordered, delalloc;
 	u64 thresh = div_factor_fine(space_info->total_bytes, 98);
 	u64 used;
 
 	/* If we're just plain full then async reclaim just slows us down. */
-	if ((space_info->bytes_used + space_info->bytes_reserved) >= thresh)
+	if ((space_info->bytes_used + space_info->bytes_reserved) +
+	     global_rsv_size >= thresh)
 		return 0;
 
 	/*

From patchwork Wed Apr 28 17:38:45 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Josef Bacik <josef@toxicpanda.com>
X-Patchwork-Id: 12229881
Return-Path: <linux-btrfs-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-16.8 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,INCLUDES_PATCH,
	MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,USER_AGENT_GIT autolearn=ham
	autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id CA029C433ED
	for <linux-btrfs@archiver.kernel.org>; Wed, 28 Apr 2021 17:40:17 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 8A02C61107
	for <linux-btrfs@archiver.kernel.org>; Wed, 28 Apr 2021 17:40:17 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S242654AbhD1RlA (ORCPT <rfc822;linux-btrfs@archiver.kernel.org>);
        Wed, 28 Apr 2021 13:41:00 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:48216 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S242747AbhD1Rjm (ORCPT
        <rfc822;linux-btrfs@vger.kernel.org>);
        Wed, 28 Apr 2021 13:39:42 -0400
Received: from mail-qk1-x730.google.com (mail-qk1-x730.google.com
 [IPv6:2607:f8b0:4864:20::730])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id 8CC1CC061574
        for <linux-btrfs@vger.kernel.org>;
 Wed, 28 Apr 2021 10:38:56 -0700 (PDT)
Received: by mail-qk1-x730.google.com with SMTP id d19so28694446qkk.12
        for <linux-btrfs@vger.kernel.org>;
 Wed, 28 Apr 2021 10:38:56 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=toxicpanda-com.20150623.gappssmtp.com; s=20150623;
        h=from:to:subject:date:message-id:in-reply-to:references:mime-version
         :content-transfer-encoding;
        bh=KHZQTqAt9jx85lePUXDJmc1Z5J9qpV+hyISF/8Horhk=;
        b=aYOA6TibqdW0nkvgcwYmFF0NTzoB1l7cgaRFx+5/12FejpMOq9+FqmBcwSW1mD/yV8
         GTf3/nkAakdtdhU9A7qfOHe781qtzyBxsJ7xJ3Ek9uDXUmfSnhUZeik+naE8EPfaSXHg
         Qn1O8b1nl7/RBOIJW4hZmMP/WzZlLMClSSMPuxcJOXm2vWO74Ptj2iurHpEKqV0rGixj
         vS57gXRIg+9Q14fPaUnp0ktwRuroGGJskb4wc7UPVemVtFBNVGxQvbYgVgV+Zp+TcfL1
         2gMPox/dmMnji8NJ0vYGSw1hq8nt4hTdIxJngEMBo61VYJYFG0QT7aX4p+EhgxB8UeT2
         iNIQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=KHZQTqAt9jx85lePUXDJmc1Z5J9qpV+hyISF/8Horhk=;
        b=EeS3mwomacBLqzEbk1eTz0PhbajlfPrNnMG4raRiAicFv1VuaKwriZHB27Ocw+Dkt1
         xoUycaD5Jaga651JB5wIKn7U6BH1KG6cQF6IcUyMO05bGD+M61jweTrz1ZSHPQQGh6wN
         fos0bfPFfwia2wgFef66yAF3TRtgEZhob/dR5dZYtpjx13DdR8K0dEw97V+sc+Rpb41/
         6UJ+NGS4aUgKpgXSJnPd8BjcIh/aFEp4VAJIBLu8jrwD6lDS2XluibaLgojfuAYpANDK
         zIp5k+7P4F77gpzOhB7I4Vopr7JEOmyO3g/SLj3EHxgulYl73I1NUxEj1qDSeBayQBsL
         Q6vA==
X-Gm-Message-State: AOAM532nLTCepiIBmkZ+Svzyn4M1rJAkgWDrDoxQ2BYBjEcgUVHQsm8r
        nKnUwF81klNSZD0TgT69m7HuDnhFTGtSnQ==
X-Google-Smtp-Source: 
 ABdhPJxHKNzCKmy6pF6isltWCUYcO5gcxIniNeWfx/Uc5G4fDXXCxcbkiwFfn3w82RODg08fOHTifA==
X-Received: by 2002:a37:b702:: with SMTP id h2mr2972326qkf.88.1619631535435;
        Wed, 28 Apr 2021 10:38:55 -0700 (PDT)
Received: from localhost (cpe-174-109-172-136.nc.res.rr.com.
 [174.109.172.136])
        by smtp.gmail.com with ESMTPSA id
 q10sm304953qke.55.2021.04.28.10.38.54
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Wed, 28 Apr 2021 10:38:55 -0700 (PDT)
From: Josef Bacik <josef@toxicpanda.com>
To: linux-btrfs@vger.kernel.org, kernel-team@fb.com
Subject: [PATCH 4/7] btrfs: use the global rsv size in the preemptive thresh
 calculation
Date: Wed, 28 Apr 2021 13:38:45 -0400
Message-Id: 
 <31114026fc63ffebcdc43197fded45da7731ef03.1619631053.git.josef@toxicpanda.com>
X-Mailer: git-send-email 2.26.3
In-Reply-To: <cover.1619631053.git.josef@toxicpanda.com>
References: <cover.1619631053.git.josef@toxicpanda.com>
MIME-Version: 1.0
Precedence: bulk
List-ID: <linux-btrfs.vger.kernel.org>
X-Mailing-List: linux-btrfs@vger.kernel.org

We calculate the amount of "free" space available for normal
reservations by taking the total space and subtracting out the hard used
space, which is readonly, used, and reserved space.  However we weren't
taking into account the global block rsv, which is essentially hard used
space.  Handle this by subtracting it from the available free space, so
that our threshold more closely mirrors reality.

Signed-off-by: Josef Bacik <josef@toxicpanda.com>
---
 fs/btrfs/space-info.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/fs/btrfs/space-info.c b/fs/btrfs/space-info.c
index e341f995a7dd..b0dd9b57d352 100644
--- a/fs/btrfs/space-info.c
+++ b/fs/btrfs/space-info.c
@@ -840,8 +840,10 @@ static bool need_preemptive_reclaim(struct btrfs_fs_info *fs_info,
 
 	thresh = calc_available_free_space(fs_info, space_info,
 					   BTRFS_RESERVE_FLUSH_ALL);
-	thresh += (space_info->total_bytes - space_info->bytes_used -
-		   space_info->bytes_reserved - space_info->bytes_readonly);
+	used = space_info->bytes_used + space_info->bytes_reserved +
+		space_info->bytes_readonly + global_rsv_size;
+	if (used < space_info->total_bytes)
+		thresh += space_info->total_bytes - used;
 	thresh >>= space_info->clamp;
 
 	used = space_info->bytes_pinned;

From patchwork Wed Apr 28 17:38:46 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Josef Bacik <josef@toxicpanda.com>
X-Patchwork-Id: 12229879
Return-Path: <linux-btrfs-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-16.8 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,INCLUDES_PATCH,
	MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,USER_AGENT_GIT autolearn=ham
	autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 340CBC43460
	for <linux-btrfs@archiver.kernel.org>; Wed, 28 Apr 2021 17:40:16 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id EF4B56143F
	for <linux-btrfs@archiver.kernel.org>; Wed, 28 Apr 2021 17:40:15 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S242648AbhD1Rk7 (ORCPT <rfc822;linux-btrfs@archiver.kernel.org>);
        Wed, 28 Apr 2021 13:40:59 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:48218 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S242763AbhD1Rjo (ORCPT
        <rfc822;linux-btrfs@vger.kernel.org>);
        Wed, 28 Apr 2021 13:39:44 -0400
Received: from mail-qk1-x736.google.com (mail-qk1-x736.google.com
 [IPv6:2607:f8b0:4864:20::736])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id C7910C0613ED
        for <linux-btrfs@vger.kernel.org>;
 Wed, 28 Apr 2021 10:38:57 -0700 (PDT)
Received: by mail-qk1-x736.google.com with SMTP id 190so7723029qkl.11
        for <linux-btrfs@vger.kernel.org>;
 Wed, 28 Apr 2021 10:38:57 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=toxicpanda-com.20150623.gappssmtp.com; s=20150623;
        h=from:to:subject:date:message-id:in-reply-to:references:mime-version
         :content-transfer-encoding;
        bh=UqetKq/bdwiqKwQ7OIwmuOIYYkvXlIOefyVfaVdNg2g=;
        b=2ShFlTkLFcEa0F0cA08PMTKOPHwxjiQ80uCWfaLY+WJWSeNr/IDLiNPtyt5GQlQIWu
         fD5gL5y9rioyYYbyguMA2G5Jvh4wJylWF9V0488Y2sZO0PsQsC3VNL/Hg3BIvejyuZuu
         wxOdjgW+q/I4tkgnsf13TQ8z4US2H/ee26PUEU0v8OrmmW9Yae4O9W8K7HYnVK9pxpng
         LCw5VTPQrrL8dhkwjylgq0sGdBg9ivoP6bJB+rE/6vm5iMubVw+Gxbmvu7QpZ3bcHlkK
         AV5OjQCIEAPKiodyJ+uorQ8uixPYN8yBIEEbysKoqh7REhgqu5MXBxOts6uBClU4royR
         vkyA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=UqetKq/bdwiqKwQ7OIwmuOIYYkvXlIOefyVfaVdNg2g=;
        b=DA6BbsZ66jAnkqhoAxAyhzJxOScSjX3ilhj/O0/Y7oKmt35/frkTmw6CXCgdMPKlRa
         ubf86HNxsQFRgk2PEMojCSEYx+x1ilL6FsMOr7lgfeF/tdyKrX7p2rsDyBgJYVabF5oc
         KC7VCR0/we6fczPmffgWOcPfojYDXXSFNR8gVL9esHB2RYvOt/RrJ/1mHC3hgJpvZgUc
         nSpQZ0MZjYThZm1xAwHtUlmWVF96ZeTV9VjwNioCQ3uwa52gEkEj3/kvPC7Umymj3ITF
         fb3TWxr63UiV7VVBC26kHymBfc99MYJxc8NqDcLdOB106qIK5Y2eqfRtkWY08UK7+Dql
         EBNw==
X-Gm-Message-State: AOAM532YfLvlaOyOdrdXGU6cfzIeU417yQvO91pBKJYcpJAFQ/k+t9kR
        CUzNAI1PbbVIf698ieMyg0FImRWzARmmOA==
X-Google-Smtp-Source: 
 ABdhPJw79KQN5m9If8ko1T+bGKVic8lx3wp2BFjwe7jjzViPco79oP7Z4N7zeTk7GqBIMez7tIyeSA==
X-Received: by 2002:a37:ad0:: with SMTP id 199mr30182088qkk.283.1619631536797;
        Wed, 28 Apr 2021 10:38:56 -0700 (PDT)
Received: from localhost (cpe-174-109-172-136.nc.res.rr.com.
 [174.109.172.136])
        by smtp.gmail.com with ESMTPSA id
 u18sm302025qku.39.2021.04.28.10.38.56
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Wed, 28 Apr 2021 10:38:56 -0700 (PDT)
From: Josef Bacik <josef@toxicpanda.com>
To: linux-btrfs@vger.kernel.org, kernel-team@fb.com
Subject: [PATCH 5/7] btrfs: don't include the global rsv size in the
 preemptive used amount
Date: Wed, 28 Apr 2021 13:38:46 -0400
Message-Id: 
 <612a917a29754cb8f2745a9d45a2d78dfee538fc.1619631053.git.josef@toxicpanda.com>
X-Mailer: git-send-email 2.26.3
In-Reply-To: <cover.1619631053.git.josef@toxicpanda.com>
References: <cover.1619631053.git.josef@toxicpanda.com>
MIME-Version: 1.0
Precedence: bulk
List-ID: <linux-btrfs.vger.kernel.org>
X-Mailing-List: linux-btrfs@vger.kernel.org

When deciding if we should preemptively flush space, we will add in the
amount of space used by all block rsvs.  However this also includes the
global block rsv, which isn't flushable so shouldn't be accounted for in
this calculation.  If we decide to use ->bytes_may_use in our used
calculation we need to subtract the global rsv size from this amount so
it most closely matches the flushable space.

Signed-off-by: Josef Bacik <josef@toxicpanda.com>
Reviewed-by: Johannes Thumshirn <johannes.thumshirn@wdc.com>
---
 fs/btrfs/space-info.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/fs/btrfs/space-info.c b/fs/btrfs/space-info.c
index b0dd9b57d352..4e3857474cfd 100644
--- a/fs/btrfs/space-info.c
+++ b/fs/btrfs/space-info.c
@@ -871,7 +871,7 @@ static bool need_preemptive_reclaim(struct btrfs_fs_info *fs_info,
 		used += fs_info->delayed_refs_rsv.reserved +
 			fs_info->delayed_block_rsv.reserved;
 	else
-		used += space_info->bytes_may_use;
+		used += space_info->bytes_may_use - global_rsv_size;
 
 	return (used >= thresh && !btrfs_fs_closing(fs_info) &&
 		!test_bit(BTRFS_FS_STATE_REMOUNTING, &fs_info->fs_state));

From patchwork Wed Apr 28 17:38:47 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Josef Bacik <josef@toxicpanda.com>
X-Patchwork-Id: 12229877
Return-Path: <linux-btrfs-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-16.8 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,INCLUDES_PATCH,
	MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,USER_AGENT_GIT autolearn=ham
	autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id A7A73C433B4
	for <linux-btrfs@archiver.kernel.org>; Wed, 28 Apr 2021 17:40:14 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 6D8EF613E7
	for <linux-btrfs@archiver.kernel.org>; Wed, 28 Apr 2021 17:40:14 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S240429AbhD1Rkz (ORCPT <rfc822;linux-btrfs@archiver.kernel.org>);
        Wed, 28 Apr 2021 13:40:55 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:48228 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S242770AbhD1Rjo (ORCPT
        <rfc822;linux-btrfs@vger.kernel.org>);
        Wed, 28 Apr 2021 13:39:44 -0400
Received: from mail-qv1-xf36.google.com (mail-qv1-xf36.google.com
 [IPv6:2607:f8b0:4864:20::f36])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id 67F9AC061573
        for <linux-btrfs@vger.kernel.org>;
 Wed, 28 Apr 2021 10:38:59 -0700 (PDT)
Received: by mail-qv1-xf36.google.com with SMTP id a30so4444165qvb.12
        for <linux-btrfs@vger.kernel.org>;
 Wed, 28 Apr 2021 10:38:59 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=toxicpanda-com.20150623.gappssmtp.com; s=20150623;
        h=from:to:subject:date:message-id:in-reply-to:references:mime-version
         :content-transfer-encoding;
        bh=W0QBmd3JB0oGkZyaFdM3QZpUQdmhZeiq7BV9Xb1MG/Y=;
        b=zJrfnhpvu5m6JwuZRwA33yBY9f0M9mP28b5V9GpjJtxu3Z1q5jIpYsHg2ginlmkCni
         U5EjOify0C73Z0Xvh7WAPbznOaz04ZCcuWGr0woMY+0272XKEGfsNtT83/xNNDBB3oWI
         gKWdAtGFKOXPU1bvGOWfeoxPY73/lbHmYBVc6B/fyJURdp9sH/hw1B+X/ZJaMAOmJpRS
         Jcs1aosnP30ucVC4IzNopdwBp9WaiRaQbupq69KXTFOxFD4q2kmN6m5NmFQY02oE1Xp+
         yn0dJHhc+a3dGSUiTRhxNi/2dpM1XbI0yQ6MU+dj4TeQsu2ezB526CNAfPFjE7G20Q1g
         G7Rg==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=W0QBmd3JB0oGkZyaFdM3QZpUQdmhZeiq7BV9Xb1MG/Y=;
        b=Lax9hbB2DmWwLJEDg4YePPavTA9MXY0kOc6Tgc/v/xlgwUaqbozD0vw/37iU9SJFs0
         w3/BmwCtww8XyTuRR6OQY8UTU3q9cG4yLvb0taoWHWOqb5COW1glsrslORQK+ZaA8rkT
         t6rhzdNLXNLXZmc9JFDo7VZHeNsM8qugKI98uwqr4BoxhG2Gxwgh2QEK1I47lz807Ejz
         mydpKKOfA3WSCLM2gvZ93FWKc39ygFWSbSqtVbuuQMfQj23AzPuKI2E+eu2y84J617b8
         3kEEviSgANx/gkt3Ye+WCk/gj5+cf/F49PQo5J6Gp94y0iF3qgLZrZ+yD/EhzOfLcdJF
         svjQ==
X-Gm-Message-State: AOAM533H1uItTBjzS69cL8uwpG5OQJqdJ1qydBzGl4Mf8zKIOGAjBVAS
        FHUJeVDXCFYS6Tm+hKZxq9x6CvH3WRey2Q==
X-Google-Smtp-Source: 
 ABdhPJxFlyBisYO+95Rh1BlBCposUylQMjOuQtvZTRE7yKluqpYYTzAAI3CLBR2Wuiz0+IfNQnFQWA==
X-Received: by 2002:a0c:e50e:: with SMTP id l14mr6624923qvm.52.1619631538369;
        Wed, 28 Apr 2021 10:38:58 -0700 (PDT)
Received: from localhost (cpe-174-109-172-136.nc.res.rr.com.
 [174.109.172.136])
        by smtp.gmail.com with ESMTPSA id
 h188sm306216qkd.23.2021.04.28.10.38.57
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Wed, 28 Apr 2021 10:38:57 -0700 (PDT)
From: Josef Bacik <josef@toxicpanda.com>
To: linux-btrfs@vger.kernel.org, kernel-team@fb.com
Subject: [PATCH 6/7] btrfs: only ignore delalloc if delalloc is much smaller
 than ordered
Date: Wed, 28 Apr 2021 13:38:47 -0400
Message-Id: 
 <c8ac5979d064d6574eb17a99110ae36023025e3e.1619631053.git.josef@toxicpanda.com>
X-Mailer: git-send-email 2.26.3
In-Reply-To: <cover.1619631053.git.josef@toxicpanda.com>
References: <cover.1619631053.git.josef@toxicpanda.com>
MIME-Version: 1.0
Precedence: bulk
List-ID: <linux-btrfs.vger.kernel.org>
X-Mailing-List: linux-btrfs@vger.kernel.org

While testing heavy delalloc workloads I noticed that sometimes we'd
just stop preemptively flushing when we had loads of delalloc available
to flush.  This is because we skip preemptive flushing if delalloc <=
ordered.  However if we start with say 4gib of delalloc, and we flush
2gib of that, we'll stop flushing there, when we still have 2gib of
delalloc to flush.

Instead adjust the ordered bytes down by half, this way if 2/3 of our
outstanding delalloc reservations are tied up by ordered extents we
don't bother preemptive flushing, as we're getting close to the state
where we need to wait on ordered extents.

Signed-off-by: Josef Bacik <josef@toxicpanda.com>
---
 fs/btrfs/space-info.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/fs/btrfs/space-info.c b/fs/btrfs/space-info.c
index 4e3857474cfd..cf09b23f3448 100644
--- a/fs/btrfs/space-info.c
+++ b/fs/btrfs/space-info.c
@@ -864,8 +864,14 @@ static bool need_preemptive_reclaim(struct btrfs_fs_info *fs_info,
 	 * clearly be heavy enough to warrant preemptive flushing.  In the case
 	 * of heavy DIO or ordered reservations, preemptive flushing will just
 	 * waste time and cause us to slow down.
+	 *
+	 * We want to make sure we truly are maxed out on ordered however, so
+	 * cut ordered in half, and if it's still higher than delalloc then we
+	 * can keep flushing.  This is to avoid the case where we start
+	 * flushing, and now delalloc == ordered and we stop preemptively
+	 * flushing when we could still have several gigs of delalloc to flush.
 	 */
-	ordered = percpu_counter_read_positive(&fs_info->ordered_bytes);
+	ordered = percpu_counter_read_positive(&fs_info->ordered_bytes) >> 1;
 	delalloc = percpu_counter_read_positive(&fs_info->delalloc_bytes);
 	if (ordered >= delalloc)
 		used += fs_info->delayed_refs_rsv.reserved +

From patchwork Wed Apr 28 17:38:48 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Josef Bacik <josef@toxicpanda.com>
X-Patchwork-Id: 12229871
Return-Path: <linux-btrfs-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-16.8 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,INCLUDES_PATCH,
	MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,USER_AGENT_GIT autolearn=ham
	autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id D4890C43461
	for <linux-btrfs@archiver.kernel.org>; Wed, 28 Apr 2021 17:39:36 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 9C54661107
	for <linux-btrfs@archiver.kernel.org>; Wed, 28 Apr 2021 17:39:36 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S242935AbhD1RkT (ORCPT <rfc822;linux-btrfs@archiver.kernel.org>);
        Wed, 28 Apr 2021 13:40:19 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:48236 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S242795AbhD1Rjq (ORCPT
        <rfc822;linux-btrfs@vger.kernel.org>);
        Wed, 28 Apr 2021 13:39:46 -0400
Received: from mail-qt1-x82a.google.com (mail-qt1-x82a.google.com
 [IPv6:2607:f8b0:4864:20::82a])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id DE058C061573
        for <linux-btrfs@vger.kernel.org>;
 Wed, 28 Apr 2021 10:39:00 -0700 (PDT)
Received: by mail-qt1-x82a.google.com with SMTP id n22so7700062qtk.9
        for <linux-btrfs@vger.kernel.org>;
 Wed, 28 Apr 2021 10:39:00 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=toxicpanda-com.20150623.gappssmtp.com; s=20150623;
        h=from:to:subject:date:message-id:in-reply-to:references:mime-version
         :content-transfer-encoding;
        bh=7g6xGjOHYNHKhLW5t0Cgmu0S4h9xleWu943+SBo0kL8=;
        b=A8q5xXGOVZXcIGwWvuuwDXuqCpPPENyKPZyfdOy3KN4hxuOawRSUPNrW+x0E2tY3gg
         GNzhywqJTwAyKYJOXTOg5cIl5UWOIaKmBej33FX5yodcd8COzUmw0blFq+pFIV5Pe3Tm
         fV7Kr7Ra2gIc7fdpJ9FzSnxTGWe7n9/mYpQjARZ9l/qFnLbwPkn8bcY51YSnKV647dk/
         SE5+HF5TtZRChHTQjfrygXBNhFM/8H6aAC8G71RTeuhqeHm8BvPfA9Mei/7Yo+McdK2B
         SJJTeONpddEJY8ng7aollBFEOVed8Zc9cAnLvB9xvl6b4B3O6P8tjUmLaKyDjlPyumID
         vK4g==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=7g6xGjOHYNHKhLW5t0Cgmu0S4h9xleWu943+SBo0kL8=;
        b=T7QmW9ECEBRdph4h3efHx2QJprh+dzRvNlm5LlJiM1EphJmOiof7mLMzQrZ3dAz0Pg
         lkMHLM4E99BnsihFw4BJb81wigYKqy2uMJ4zW3hGlM+TMYY5jCp3016wbRDDy24mM9vs
         D4lR6sz5NToMGfpK1OjxffGvyky7gNzldbHpIyAEI4jqQqyc4iO3fLwJSN7d7jelM60Z
         JXReZYHsGUE3rPkp/MLj8lCMOACjkFZCIgSTsXOfwOXeHQtePIxr4sM1T/5Oj/2kDHlY
         5yspllGoVcUvsjT1lV+ApxeO+y5pob23srzDvoaCvHA43p7m+EY3yvObuKKErmOxOcQs
         MoEA==
X-Gm-Message-State: AOAM531GkfUfYET/p4BViE10WLXCUUUoXgsJuCM/f1od+9Wl+B3Ql6jD
        jamtaldMzHDmT+mkdO2DSP6L85LYBRSnZw==
X-Google-Smtp-Source: 
 ABdhPJw+X3R31ueDdtX5+IgJkSJzo9H6ZZcQljDZIlRigipKDgWUadPt0fSnp09YhcwBAzYiMMkzeQ==
X-Received: by 2002:a05:622a:253:: with SMTP id
 c19mr28125158qtx.53.1619631539799;
        Wed, 28 Apr 2021 10:38:59 -0700 (PDT)
Received: from localhost (cpe-174-109-172-136.nc.res.rr.com.
 [174.109.172.136])
        by smtp.gmail.com with ESMTPSA id
 y6sm271570qkd.106.2021.04.28.10.38.59
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Wed, 28 Apr 2021 10:38:59 -0700 (PDT)
From: Josef Bacik <josef@toxicpanda.com>
To: linux-btrfs@vger.kernel.org, kernel-team@fb.com
Subject: [PATCH 7/7] btrfs: handle preemptive delalloc flushing slightly
 differently
Date: Wed, 28 Apr 2021 13:38:48 -0400
Message-Id: 
 <63c38e3b1c4b1214859852dbc83dac9115a2f26a.1619631053.git.josef@toxicpanda.com>
X-Mailer: git-send-email 2.26.3
In-Reply-To: <cover.1619631053.git.josef@toxicpanda.com>
References: <cover.1619631053.git.josef@toxicpanda.com>
MIME-Version: 1.0
Precedence: bulk
List-ID: <linux-btrfs.vger.kernel.org>
X-Mailing-List: linux-btrfs@vger.kernel.org

If we decide to flush delalloc from the preemptive flusher, we really do
not want to wait on ordered extents, as it gains us nothing.  However
there was logic to go ahead and wait on ordered extents if there was
more ordered bytes than delalloc bytes.  We do not want this behavior,
so pass through whether this flushing is for preemption, and do not wait
for ordered extents if that's the case.  Also break out of the shrink
loop after the first flushing, as we just want to one shot shrink
delalloc.

Signed-off-by: Josef Bacik <josef@toxicpanda.com>
---
 fs/btrfs/space-info.c | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/fs/btrfs/space-info.c b/fs/btrfs/space-info.c
index cf09b23f3448..b2d834b92811 100644
--- a/fs/btrfs/space-info.c
+++ b/fs/btrfs/space-info.c
@@ -495,7 +495,8 @@ static inline u64 calc_reclaim_items_nr(struct btrfs_fs_info *fs_info,
  */
 static void shrink_delalloc(struct btrfs_fs_info *fs_info,
 			    struct btrfs_space_info *space_info,
-			    u64 to_reclaim, bool wait_ordered)
+			    u64 to_reclaim, bool wait_ordered,
+			    bool for_preempt)
 {
 	struct btrfs_trans_handle *trans;
 	u64 delalloc_bytes;
@@ -532,7 +533,7 @@ static void shrink_delalloc(struct btrfs_fs_info *fs_info,
 	 * ordered extents, otherwise we'll waste time trying to flush delalloc
 	 * that likely won't give us the space back we need.
 	 */
-	if (ordered_bytes > delalloc_bytes)
+	if (ordered_bytes > delalloc_bytes && !for_preempt)
 		wait_ordered = true;
 
 	loops = 0;
@@ -551,6 +552,14 @@ static void shrink_delalloc(struct btrfs_fs_info *fs_info,
 				break;
 		}
 
+		/*
+		 * If we are for preemption we just want a one-shot of delalloc
+		 * flushing so we can stop flushing if we decide we don't need
+		 * to anymore.
+		 */
+		if (for_preempt)
+			break;
+
 		spin_lock(&space_info->lock);
 		if (list_empty(&space_info->tickets) &&
 		    list_empty(&space_info->priority_tickets)) {
@@ -702,7 +711,7 @@ static void flush_space(struct btrfs_fs_info *fs_info,
 	case FLUSH_DELALLOC:
 	case FLUSH_DELALLOC_WAIT:
 		shrink_delalloc(fs_info, space_info, num_bytes,
-				state == FLUSH_DELALLOC_WAIT);
+				state == FLUSH_DELALLOC_WAIT, for_preempt);
 		break;
 	case FLUSH_DELAYED_REFS_NR:
 	case FLUSH_DELAYED_REFS:
